### ISSUE TITLE
Should we add ourselves to the peers for an infohash we've announced?

### ISSUE THREAD

**Issue by cjb** [feross/bittorrent-dht#59]
tl;dr When we receive a `get_peers` for an infohash we have announced, should we include ourselves as a peer?  Currently we don't, and I'm not sure whether it's a bug.

Long version:

I've noticed that when one peer receives a `get_peers` for an infohash it published itself, it returns the K closest nodes instead of itself as a peer.  So the upshot is that the only way an announce works is if:

node A announces
node B hears the announce
node C calls get_peers, and node A is returned to node C by node B

Whereas if only two nodes were online, it'd be:

node A announces
node C starts up, calls get_peers, node A receives the get_peers but does not include itself as a peer, so the download can not proceed

Any idea if this is a bug, vs. me misunderstanding the protocol?  If it's a bug, is the fix to call `_addPeer()` for ourselves when we announce?

**Comment by feross** [comment]
In general, DHT nodes do not need to add an entry to their local table when they announce because a DHT is only responsible for tracking torrents with infohashes that are "near" their own. In the case of the broader bittorrent network, this would almost never happen.

However, if you're using this module to create your own DHT network, or just running tests/simulations, then you could easily run into this situation. In fact, it's why [this special case](https://github.com/feross/bittorrent-dht/blob/ae1ec0d3a1f720cd2bcc96040efebfcf7f2877e6/test/horde.js#L66-L74) was needed in the tests. I'll attempt to make a fix.

**Comment by feross** [comment]
Alrighty, this is fixed now! I add the peer that controls the DHT node to the local table, including both local and public ips. This is released as 3.0.2.

Note: the addresses will *always* be stored in the local table, even when it's not "near" our the DHT's node id, since https://github.com/feross/bittorrent-dht/issues/46 is not implemented yet.

**Comment by cjb** [comment]
Amazing, thanks so much!  I wasn't expecting that you'd write the code for me, I really appreciate it.  :)  I just tested on a live network, works great.
